pkg/provider/consulcatalog/config.go

```diff
@@ -85,6 +104,94 @@ func (p *Provider) buildConfiguration(ctx context.Context, items []itemData) *dy
 	return provider.Merge(ctx, configurations)
 }
 
+func connectTransportName(n string) string {
+	return "connect-tls-" + n
+}
+
+type connectCert struct {
+	service string
+	root    []string
+	leaf    keyPair
+}
+
+func (c *connectCert) getRoot() []tls.FileOrContent {
+	var result []tls.FileOrContent
+	for _, r := range c.root {
+		result = append(result, tls.FileOrContent(r))
+	}
+	return result
+}
+
+func (c *connectCert) getLeaf() tls.Certificate {
+	return tls.Certificate{
+		CertFile: tls.FileOrContent(c.leaf.cert),
+		KeyFile:  tls.FileOrContent(c.leaf.key),
+	}
+}
+
+func (c *connectCert) serverTransport(item itemData) *dynamic.ServersTransport {
+	sname := connectTransportName(item.Name)
+	return &dynamic.ServersTransport{
+		ServerName:         sname,
+		InsecureSkipVerify: true,
+		RootCAs:            c.getRoot(),
+		Certificates: tls.Certificates{
+			c.getLeaf(),
+		},
+		VerifyPeerCertificate: func(cfg *gtls.Config, rawCerts [][]byte, verifiedChains [][]*x509.Certificate) error {
```

Comment:
Fair point, but I'd like to leave this as is for now till we know how to continue with https://github.com/traefik/traefik/issues/7826 (which should make this unneeded completely).

Reply:
So you're saying you're waiting for some general feedback from Blake before moving forward on this PR, or are you waiting for something else?

I'm asking, because I think we kind of changed our minds here about being generic about ServersTransport and VerifyPeerCertificate, i.e. we're more inclined to make the least amount of changes to the config to make things work for Consul Connect. And we'll worry about exposing a generic mechanism for all providers later, if the need ever arises. In other words, as far as this PR is concerned I think you can forget about #7826 for now.

Reply:
I tried copying `clientSideVerifier` & `verifyServerCertMatchesURI` but I do not think the code got any cleaner, it would now parse the certificates twice & also requires extra structs to provide the proper URIs. I think I will just add some comments to clarify what is copied from where and then we maybe can go with that?

Reply:
Managed to get it to work, commit incoming.